django/utils/encoding.py

```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+# List of byte values that uri_to_iri is going to decode from percent encoding.
+# These are the unreserved characters from RFC 3986 and everything above 128,
+# because bytes >= 128 are part of multibyte unicode characters.
+_decode_ranges = [[45, 46, 95, 126], range(65, 91), range(97, 123), range(128, 256)]
+_hextobyte = {
+    (fmt % char).encode(): six.int2byte(char)
+    for range in _decode_ranges for char in range for fmt in ['%02x', '%02X']
+}
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
```

Comment:
No, that is step 3:
> `3.` Re-percent-encode any octet produced in step 2 that is not part of a strictly legal UTF-8 octet sequence.

I was confused by step 4 at first too, but it is clarified in the referenced sections:
> `4.` Re-percent-encode all octets produced in step 3 that in UTF-8 represent characters that are not appropriate according to sections 2.2, 4.1, and 6.1.

Section 2.2 says:
> […] except that the range of unreserved characters is expanded to include UCS characters, *with the restriction that private UCS characters can occur only in query parts*

Section 4.1 says:
> IRIs MUST NOT contain bidirectional formatting characters (LRM, RLM, LRE, RLE, LRO, RLO, and PDF).

Section 6.1 says:
> The UCS contains many areas of characters for which there are strong visual look-alikes.  Because of the likelihood of        transcription errors, these also should be avoided.  This        includes the full-width equivalents of Latin characters,        half-width Katakana characters for Japanese, and many others.  It        also includes many look-alikes of "space", "delims", and       "unwise", characters excluded in [RFC3491].

We check none of the above.